game-core/src/main/java/games/strategy/engine/data/util/BreadthFirstSearch.java

```diff
@@ -0,0 +1,90 @@
+package games.strategy.engine.data.util;
+
+import games.strategy.engine.data.GameMap;
+import games.strategy.engine.data.Territory;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayDeque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.ObjectUtils;
+
+/**
+ * Implements Breadth First Search (BFS) to traverse / find territories. Since the search criteria
+ * varies depending on the use case, the class is designed to be sub-classed, with methods visit()
+ * and shouldContinueSearch() that can be overridden to customize the behavior.
+ */
+public class BreadthFirstSearch {
+  private final GameMap map;
+  @Getter private final Set<Territory> visited;
+  private final ArrayDeque<Territory> territoriesToCheck;
+  private final Predicate<Territory> cond;
+
+  public BreadthFirstSearch(final Territory startTerritory, final Predicate<Territory> cond) {
+    this.map = startTerritory.getData().getMap();
+    this.visited = new HashSet<Territory>(List.of(startTerritory));
+    this.territoriesToCheck = new ArrayDeque<Territory>(List.of(startTerritory));
+    this.cond = cond;
+  }
+
+  public BreadthFirstSearch(final Territory startTerritory) {
+    this(startTerritory, Matches.always());
+  }
+
+  /**
+   * Called when a new territory is encountered. Can be overridden to provide custom search
+   * behavior.
+   *
+   * @param territory The new territory.
+   */
+  public void visit(final Territory territory) {}
+
+  /**
+   * Called after all territories within the specified distance have been searched. Can be
+   * overridden to terminate the search.
+   *
+   * @param distanceSearched The current distance searched
+   * @return Whether the search should continue.
+   */
+  public boolean shouldContinueSearch(final int distanceSearched) {
+    return true;
+  }
+
+  /**
+   * Performs the search. It will end when either all territories have been visited or
+   * shouldContinueSearch() returns false.
+   */
+  public void search() {
```

Comment:
Don't see a good way of reducing further. Note: Overall it's still a reduction since the function where this code was originally had higher complexity.

Reply:
FWIW, this one was fixed:

> Method findNearbyEnemyCapitalsAndFactories has a Cognitive Complexity of 16 (exceeds 5 allowed). Consider refactoring.
Fixed

Going from 16 -> 6 is certainly the right direction :+1: